### ISSUE TITLE
HLSL: UAV counters always in descriptor set 0

### ISSUE THREAD

**Issue by chaoticbob** [KhronosGroup/glslang#1100]
It looks like glslang always puts the UAV counters in ```DescriptorSet 0```. @danginsburg and I chatted about this briefly and were wondering if they should remain in the same descriptor set as their respective UAV buffers.

Sample shader:
```C
struct Data {
  float4 v;
};
ConsumeStructuredBuffer<Data> buf_in : register(u0, space1);
AppendStructuredBuffer<Data>  buf_out : register(u2, space3);

float4 main(float4 sv_pos : SV_POSITION) : SV_TARGET
{
  Data val = buf_in.Consume();       
  buf_out.Append(val);
  return sv_pos;
}
````

...produces...
````
                              Decorate 18(buf_in) BufferBlock
                              Decorate 20(buf_in) DescriptorSet 1
                              Decorate 20(buf_in) Binding 0
                              MemberDecorate 23(buf_in@count) 0 Offset 0
                              Decorate 23(buf_in@count) BufferBlock
                              Decorate 25(buf_in@count) DescriptorSet 0
                              Decorate 25(buf_in@count) Binding 0
                              Decorate 39(buf_out) DescriptorSet 3
                              Decorate 39(buf_out) Binding 2
                              Decorate 40(buf_out@count) DescriptorSet 0
                              Decorate 40(buf_out@count) Binding 1
````

```buf_in@count``` and ```buf_out@count``` are both in ```DescriptorSet 0``` while ```buf_in``` and ```buf_out``` are in  ```DescriptorSet 2``` and ```DescriptorSet 3```, respectively.

**Comment by LoopDawg** [comment]
That seems sensible to me, although it's down to a matter of convention rather than definition.

How does this interact with SB type sharing?  A while ago -- and unfortunately I can't immediately locate the issue number -- we made an explicit change to share the underlying SPIR-V type between different structuredbuffer objects of the same template type.  I think (?) that was to allow multiple HLSL level SB objects to operate on the same underlying buffer, or something like that.  @danginsburg, do you recall more?

I also wonder, depending on the answer to the above, whether there may be lurking defects in the way that type sharing is happening.

**Comment by johnkslang** [comment]
Independent of type, I am sure we want an object and its corresponding counter in the same descriptor set, because of how the API will use descriptor sets.

**Comment by danginsburg** [comment]
>  @danginsburg, do you recall more?

No, sorry, I actually have no recollection of this at all.  I'm not sure it was for one of our shaders or one of our issues.

> Independent of type, I am sure we want an object and its corresponding counter in the same descriptor set, because of how the API will use descriptor sets.

Yes, that's right.  I can't really think of why an app would place a UAV in one descriptor set explicitly and then want the counter in another one.  If that was really a desired behavior it would probably be better to introduce a layout qualifier where you can declare the set of the hidden counter buffer to be different than the UAV, but the default behavior I think should be for them to be in the same set.

To be clear, I think such a layout qualifier is probably unnecessary and should probably only be done if someone asks for it.

**Comment by LoopDawg** [comment]
OK, I will fix this.

**Comment by johnkslang** [comment]
@LoopDawg was this still something you are looking at?

**Comment by LoopDawg** [comment]
Yes, sorry for the delay.

#1179 is in progress.  The counter buffers now inherit the descriptor set from their parent buffer.  The PR isn't quite ready for prime time yet due to some subtleties I need to investigate around timing of parsing layout descriptors vs type sharing, but AFAICS it handles the above example already.  I'll get it finished soon.
